### PR TITLE
WIP: initial ID.me log in port for local dev environment

### PR DESCRIPTION
**Pull request by shauni**
I'll be adding tests, and have made notes on things that will need to change soon (possibly as part of this PR), but would love any comments in the mean time.

### CODE REVIEW

File: app/controllers/v0/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    def saml_callback
+      saml_response = OneLogin::RubySaml::Response.new(
+          params[:SAMLResponse], settings: SAML::SETTINGS)
+
+      if saml_response.is_valid?
+        session[:user] = {
+            name: saml_response.name_id,
+            attributes: saml_response.attributes.all.to_h
+          }
+        redirect_to v0_welcome_path
```
Comment (shauni) [inline app/controllers/v0/sessions_controller.rb]: TODO: redirect to original url instead of default

File: config/initializers/saml_settings.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
+module SAML
```
Comment (shauni) [inline config/initializers/saml_settings.rb]: Lots of changes to come here in follow-on PRs, once we figure out the devops story a bit better

File: app/controllers/v0/sessions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            uid: attributes['uuid']
+          }
+
+        render json: profile
+      else
+        redirect_to action: "create"
+      end
+    end
+
+    def destroy
```
Comment (shauni) [inline app/controllers/v0/sessions_controller.rb]: need to actually log out of ID.me also

File: Gemfile [hunk 4]
```diff
@@ -28,6 +28,8 @@ gem 'sdoc', '~> 0.4.0', group: :doc
 # Discussion: https://github.com/18F/college-choice/issues/597#issuecomment-139034834
 gem "puma", "~> 2.16.0"
 
+gem 'ruby-saml'
```
Comment (robbiethegeek) [inline Gemfile]: should we lock this to a version?

**Comment by robbiethegeek** [comment]
Any idea on a testing story for this code? is there a pattern we can apply to test this?

File: app/controllers/v0/example_controller.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+module V0
+  class ExampleController < ApplicationController
+
+    def index
+      render json: { "message": "Welcome to the vets.gov API"}
+    end
+
+    # requires log in
+    def welcome
+      if session[:user]
```
Comment (kreek) [inline app/controllers/v0/example_controller.rb]: Is this conditional for demo purposes? Could it become a devise like before_action?

File: Gemfile (hunk 4 above)
Comment (kreek) [inline Gemfile]: +1 with ~>

File: app/controllers/v0/example_controller.rb (hunk 5 above)
Comment (shauni) [inline app/controllers/v0/example_controller.rb]: yes! I meant make it into a before_action, will fix :)

File: .scss-lint.yml [hunk 6]
```diff
@@ -1,147 +0,0 @@
-scss_files: 'app/assets/stylesheets/**/*.scss'
```
Comment (saneshark) [inline .scss-lint.yml]: should these changes be in a separate PR? There is a lot in the roadrunner app I would like for us to eventually change (ie. I couldn't help but notice that one of the gems adds Sinatra to our Gemfile.lock as dependency, that gem should probably be replaced by something else that doesn't do that for example.) 

But this change seems like it should be a in a separate PR.

File: Gemfile.lock [hunk 7]
```diff
@@ -1,13 +1,3 @@
-GIT
```
Comment (saneshark) [inline Gemfile.lock]: hah, here is the gem with sinatra. :) Still wondering if maybe these changes should be in a separate PR?

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -4,8 +4,18 @@ class ApplicationController < ActionController::Base
...
+  private
+
   def set_app_info_headers
     headers["X-GitHub-Repository"] = "https://github.com/department-of-veterans-affairs/vets-api"
     headers["X-Git-SHA"] = AppInfo::GIT_REVISION
   end
+
+  def require_login
+    unless session[:user]
+      flash[:after_login_controller] = request.parameters["controller"]
```
Comment (saneshark) [inline app/controllers/application_controller.rb]: The flash uses ActionDispatch::Session::CacheStore, which is a cookie. I think this solution is just-for-now, but perhaps it would be good to be a TODO: in here to suggest that this will do something with JSON instead. We probably don't want to use flash or session cookies in this app.

Comment (saneshark) [inline app/controllers/application_controller.rb]: Perhaps, rather than a redirect, we should raise an exception, and then have that exception caught in application controller as a particular type of exception, with JSON response body. APIs and redirects don't work well.

File: config/initializers/saml_settings.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+  SETTINGS = OneLogin::RubySaml::Settings.new
+
+  # --- Once we have a staging and production config, these will be environment specific
+  # we may turn off ID.me in dev mode
+  SETTINGS.certificate                    = File.read('../certs/shauni-server.crt')
+  SETTINGS.private_key                    = File.read('../certs/shauni-server.key')
+  SETTINGS.issuer                         = "saml-rp.shauni.localhost"
```
Comment (greggersh) [inline config/initializers/saml_settings.rb]: Suggest these be env vars, but it looks like this is in flux now.

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -4,8 +4,18 @@ class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
   before_action :set_app_info_headers
 
+  private
+
   def set_app_info_headers
     headers["X-GitHub-Repository"] = "https://github.com/department-of-veterans-affairs/vets-api"
```
Comment (saneshark) [inline app/controllers/application_controller.rb]: So I'm only raising this one out of curiosity, what exactly is this for and why is it necessary for the header? Is this a roadrunner artifact? I think we should look at calling it Vets Api V1 or something, and use numbered versioning vs hashed based X-Git-SHA for the revision?

File: config/routes.rb [hunk 11]
```diff
@@ -1,7 +1,14 @@
 Rails.application.routes.draw do
 
+  get '/auth/saml/callback', to: 'v0/sessions#saml_callback', module: 'v0'
+
   namespace :v0, defaults: {format: 'json'} do
+    resource :sessions
```
Comment (saneshark) [inline config/routes.rb]: Should we maybe put an only: [<routes to be exposed>] for resource :sessions ?

File: spec/spec_helper.rb [hunk 12]
```diff
@@ -1,6 +1,9 @@
 require "simplecov"
+
 SimpleCov.start do
   add_filter "initializers/git_revision"
+  add_filter "v0/example_controller"
+  add_filter "v0/sessions_controller"
```
Comment (saneshark) [inline spec/spec_helper.rb]: if we're going to put these filters in to prevent simplecov from picking up these files for coverage **temporarily** maybe we should put a TODO: or FIXME: comment in there as well so we know to remove this in the future, or else we might end up thinking we have 90% coverage in the future when we don't? Just a nitpick as I know this PR is just a WIP to get us moving along.

File: app/controllers/application_controller.rb (hunk 10 above)
Comment (robbiethegeek) [inline app/controllers/application_controller.rb]: it was a thought for determining the project before we talked about single repo, don't think this is a problem to include
